neuralmonkey/encoders/factored_encoder.py

```diff
@@ -1,201 +1,202 @@
+                             "lists need to have the same length")
+
+        if max_input_len is not None and max_input_len <= 0:
+            raise ValueError("Input length must be a positive integer.")
+
+        if any([esize <= 0 for esize in embedding_sizes]):
+            raise ValueError("Embedding size must be a positive integer.")
+
+        if rnn_size <= 0:
+            raise ValueError("RNN size must be a positive integer.")
+
+        if not issubclass(self.rnn_cell, tf.contrib.rnn.RNNCell):
+            raise TypeError("RNN cell must be a subclass of "
+                            "tf.contrib.rnn.RNNCell")
+    # pylint: enable=too-many-arguments,too-many-locals
+
+    @tensor
+    def input_factors(self) -> Dict[str, tf.Tensor]:
+        return {name: tf.placeholder(tf.int32, [None, None],
+                                     "encoder_factor_{}".format(name))
+                for name in self.data_ids}
+
+    # pylint: disable=no-self-use
+    @tensor
+    def input_mask(self) -> tf.Tensor:
+        return tf.placeholder(tf.float32, [None, None], "encoder_padding")
+
+    @tensor
+    def train_mode(self) -> tf.Tensor:
+        return tf.placeholder(tf.bool, [], "train_mode")
+    # pylint: enable=no-self-use
+
+    @tensor
+    def embedding_matrices(self) -> Dict[str, tf.Tensor]:
+        # NOTE the note from the decoder's embedding matrix function applies
+        # here also:
+        with tf.variable_scope("input_projection"):
+            return {name:
+                    tf.get_variable(
+                        "embeddings_factor_{}".format(name),
+                        [voc_size, emb_size],
+                        initializer=tf.random_normal_initializer(stddev=0.01))
+                    for name, emb_size, voc_size
+                    in zip(self.data_ids,
+                           self.embedding_sizes,
+                           self.vocabulary_sizes)}
+
+    @tensor
+    def bidirectional_rnn(self) -> Tuple[Tuple[tf.Tensor, tf.Tensor],
+                                         Tuple[tf.Tensor, tf.Tensor]]:
+        embedded_factors = []
+        for name in self.data_ids:
+            # pylint: disable=unsubscriptable-object
+            embedded_f = tf.nn.embedding_lookup(self.embedding_matrices[name],
+                                                self.input_factors[name])
+            # pylint: enable=unsubscriptable-object
+
+            embedded_f = dropout(embedded_f, self.dropout_keep_prob,
+                                 self.train_mode)
+            embedded_factors.append(embedded_f)
+
+        birnn_input = tf.concat(embedded_factors, 2)
+        sequence_lengths = tf.to_int32(tf.reduce_sum(self.input_mask, 1))
+
+        fw_cell, bw_cell = self._rnn_cells()  # type: RNNCellTuple
+        return tf.nn.bidirectional_dynamic_rnn(
+            fw_cell, bw_cell, birnn_input, sequence_length=sequence_lengths,
+            dtype=tf.float32)
+
+    @tensor
+    def hidden_states(self) -> tf.Tensor:
```

Comment:
jo no to bych nechal až do dalšího PR, kterej sjednotí ty enkodéry do jednoho. Zatim mi šlo o jednotný apíčko factored enkoderu podle sentence..